### ISSUE TITLE
Indirect GLX not functioning in some cases

### ISSUE THREAD

**Issue by bbarbarab** [XQuartz/XQuartz#144]
Hi,
I'm using a Qt application (running on Ubuntu 20.04, but it's the same with Ubuntu 18.04) with a login module at start.
On the macos (Catalina) I always get the known message "libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast".
I see the login module, but I've to wait for a (long) while before being able to digit inside the user and password fields.
Once it's done a first time, then the next times it works immediately.
Any suggestions about what to do on mac or Ubuntu side?
Thank you

**Comment by DrGo** [comment]
I see the same errors with some app (eg alacritty) on Big Sur 11.2.3. XQuartz 2.8.0 (xorg-server 1.19.7)
I have set  "enable_iglx" = 1 but I am still getting the errors. I say the same error with firefox but it is still usable. Alacritty on the other hand just shows empty distorted windows, but it works fine on the remote box (archlinux). 
![image](https://user-images.githubusercontent.com/12840815/113379050-1edd3100-933e-11eb-998e-b08b1a586939.png)

~> LIBGL_DEBUG=verbose glxinfo > glxinfo.log
libGL: MESA-LOADER: dlopen(/usr/lib/dri/swrast_dri.so)
libGL: Can't open configuration file /etc/drirc: No such file or directory.
libGL: Can't open configuration file /home/salah/.drirc: No such file or directory.
libGL: Can't open configuration file /etc/drirc: No such file or directory.
libGL: Can't open configuration file /home/salah/.drirc: No such file or directory.
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast

**Comment by jeremyhu** [comment]
This sounds similar to what folks were reporting and they said was addressed by #54.  However, I had doubts about whether the fix in that ticket would actually do anything.

**Comment by jeremyhu** [comment]
Can you please identify if the issue is with the clinet libraries or the server?  To do so:

  1. install 2.7.11
  2. backup XQuartz.app (I rename it as XQuartz-2.7.11.app and create a tarball of it just in case I need it later)
  3. install 2.8.0
  4. Launch your (old) Xquartz-2.7.11.app and use it to connect to your remote system.  Do you see an issue?  If so, this means it's an X11 server regression.  If not, it means it's probably an isuse in mesa or some other client library.

**Comment by DrGo** [comment]
2.7.11 and 2.8.0 gave the same errors.

**Comment by bbarbarab** [comment]
I tried (I think) all XQuartz versions since 2.7 to 2.8 and got the same behaviour. Setting enable_iglx to true had no impact.

**Comment by jeremyhu** [comment]
Oh, ok.  So this isn't a regression with 2.8.0, right?  I misunderstood that.  Ok.

**Comment by dillera** [comment]
GLX works in no cases I can see.

**Comment by jontis** [comment]
Adding to this. 
Big Sur and QQuartz 2.8.1.
ssh -X or -Y
runs xeyes and xclock properly.
glxgears does not run.
glxinfo:
ibGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  149 (GLX)
  Minor opcode of failed request:  6 (X_GLXIsDirect)
  Serial number of failed request:  37
  Current serial number in output stream:  36

**Comment by jeremyhu** [comment]
@jontis Did you actually enable IGLX though?

    defaults write org.xquartz.X11 enable_iglx -bool

**Comment by dillera** [comment]
i get 149 (GLX) failed optcode...

dillera@fuel ~ $ glxinfo 
display: localhost:10.0
server glx vendor string: SGI
server glx version string: 1.4
server glx extensions (GLX_):
    ARB_multisample, EXT_import_context, EXT_visual_info, EXT_visual_rating,
    OML_swap_method, SGI_make_current_read, SGIS_multisample, SGIX_fbconfig,
    SGIX_pbuffer, SGIX_visual_select_group.
client glx version 1.3
client glx extensions (GLX_):
    EXT_import_context, EXT_visual_info, EXT_visual_rating,
    SGI_make_current_read, SGIS_multisample, SGIX_fbconfig, SGIX_pbuffer.
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  149 (GLX)
  Minor opcode of failed request:  3 (X_GLXCreateContext)
  Value in failed request:  0x0
  Serial number of failed request:  20
  Current serial number in output stream:  23

**Comment by HeroCC** [comment]
I'm able to get steps to reproduce on my end, XQuartz 2.8.1. These same steps on a different machine running XQuartz 2.7.11 run correctly. This leads me to believe it is a regression, though I haven't done the steps @jeremyhu suggested to confirm yet.

```bash
# Ensure IGLX is enabled 
$ defaults read org.xquartz.X11 enable_iglx
1

# Quit out of the XQuartz app

# Start XQuartz and allow access from all networked clients (you also need to enable this in XQuartz security settings)
$ xhost +
access control disabled, clients can connect from any host

# Run glxgears (in a docker container for ease of install & reproduction)
$ docker run --rm -it -v /tmp/.X11-unix:/tmp/.X11-unix -e DISPLAY=host.docker.internal:0 tbillah/glxgears-docker:latest glxgears
4610 frames in 6.9 seconds = 667.255 FPS
12192 frames in 5.0 seconds = 2432.797 FPS
25512 frames in 5.0 seconds = 5102.387 FPS
# You should see the gears open on the desktop. Wait a few seconds, then...
# Click the X on the glxgears window, but leave the XQuartz app open. The terminal will return to the prompt
X connection to host.docker.internal:0 broken (explicit kill or server shutdown).

# Now, run the same command again
$ docker run --rm -it -v /tmp/.X11-unix:/tmp/.X11-unix -e DISPLAY=host.docker.internal:0 tbillah/glxgears-docker:latest glxgears
X Error of failed request:  BadAlloc (insufficient resources for operation)
  Major opcode of failed request:  149 (GLX)
  Minor opcode of failed request:  3 (X_GLXCreateContext)
  Serial number of failed request:  25
  Current serial number in output stream:  26

# Failed to launch :(
```

**Comment by jeremyhu** [comment]
Thanks.  That points to maybe some resource leak in the server.  Unfortunately, I don't really have much time to triage this at the moment (and IGLX is super-duper deprecated and unsupported these days in general), but the above data is great and will be useful when I have some time to circle back to this in the future.

**Comment by kegsay** [comment]
Running 2.8.1 and restarting XQuartz does indeed help. Similarly, I'm running a docker container which has a GUI and am using XQuartz to make it work. First time works, subsequent times do not and fail with:
```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X11 Error: GLXBadContext
X11 Error: BadAlloc (insufficient resources for operation)
```

Initially I thought this was a Docker thing so I bumped up the memory on the container, but this didn't help.

**Comment by wright-cemrc-projects** [comment]
We are having a problem with Big Sur 11.4 and XQuartz.  
It appears that after an update of Big Sur, that XQuartz support of the IGLX extension breaks. We haven't found a combination of XQuartz from 2.7.11 to 2.8.1 versions that will work with Big Sur 11.4 and have the IGLX extension working with a remote ssh connection requiring IGLX.

We are using ssh -YC to connect to the remote host, XQuartz options 'Security->Authenticate connections, and Allow connections from network clients' are both enabled.

`defaults write org.macosforge.xquartz.X11 enable_iglx -bool true`

and 

`defaults write org.xquartz.X11 enable_iglx -bool true`

Are both set, and can be checked to have a value of 1.  

It appears that after an OSX update or reboot that some part of XQuartz is getting broken and the IGLX extension no longer works on this MacBook. We've been using XQuartz 2.8.1 on other Big Sur MacBooks without issue.

We've "fixed" this issue by reinstalling XQuartz, setting values again, and trying. However in this latest event we are unable to find a combination of updating XQuartz and settings that is able to get the IGLX extension working.

Any ideas for what could be needed for XQuartz + how to prevent OSX updates or changes from breaking XQuartz IGLX support for remote connections?  Is there a version of XQuartz we should use?

**Comment by larsonmattr** [comment]
We are having a problem with Big Sur 11.4 and XQuartz.
It appears that after an update of Big Sur, that XQuartz support of the IGLX extension breaks. We haven't found a combination of XQuartz from 2.7.11 to 2.8.1 versions that will work with Big Sur 11.4 and have the IGLX extension working with a remote ssh connection requiring IGLX.

We are using ssh -YC to connect to the remote host, XQuartz options 'Security->Authenticate connections, and Allow connections from network clients' are both enabled.

defaults write org.macosforge.xquartz.X11 enable_iglx -bool true

and

defaults write org.xquartz.X11 enable_iglx -bool true

Are both set, and can be checked to have a value of 1.

It appears that after an OSX update or reboot that some part of XQuartz is getting broken and the IGLX extension no longer works on this MacBook. We've been using XQuartz 2.8.1 on other Big Sur MacBooks without issue.

We've "fixed" this issue by reinstalling XQuartz, setting values again, and trying. However in this latest event we are unable to find a combination of updating XQuartz and settings that is able to get the IGLX extension working.

Any ideas for what could be needed for XQuartz + how to prevent OSX updates or changes from breaking XQuartz IGLX support for remote connections? Is there a version of XQuartz we should use?

**Comment by larsonmattr** [comment]
Update, that the 2.8.1 with IGLX started working after a reboot of the MacBook Pro.  Still we are having persisting issues where the IGLX stops working after OSX updates or other events. Any ideas would be helpful to fix this.

**Comment by salinafali** [comment]
Having same issue on XQuartz 2.8.0 and 2.8.1. glxgears and glxinfo work fine locally but not in remote connections

**Comment by larsonmattr** [comment]
The "fix" so far on this issue with XQuartz and OSX Big Sur has been to restart the MacBook Pro, and afterwards XQuartz 2.8.1 with the GLX extension will work again on remote sessions. Maybe there is a process of XQuartz that is remaining in memory and is preventing subsequent sessions from working - which gets cleared on the restart?

**Comment by ethaniel** [comment]
> libGL error: No matching fbConfigs or visuals found
> libGL error: failed to load driver: swrast

in my case these 2 errors were solved with `sudo apt -y install libnvidia-gl-460`

**Comment by larsonmattr** [comment]
I can also note that we have issues on Linux servers with NVIDIA cards, when using the indirect rendering and MacOSX XQuartz clients.

The GLX extension fails with the Major opcode of failed request: 149 (GLX)
Minor opcode of failed request: 5 (X_GLXMakeCurrent) on a newer driver.

NVIDIA drivers 495.29.05-1 will not work with GLX and remote indirect rendering with XQuartz.
NVIDIA drivers 465.19.01-1 work with GLX and remote indirect rendering with XQuartz.  

Confirmed this on multiple systems, different GPUs.

**Comment by jewes** [comment]
```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  149 (GLX)
  Minor opcode of failed request:  24 (X_GLXCreateNewContext)
  Value in failed request:  0x0
  Serial number of failed request:  27
  Current serial number in output stream:  28
```

**Comment by kmiklas** [comment]
+1 on this error: Trying to redirect google-chrome running on Ubuntu 20.04, EC2 container to Mac Big Sur with xquartz 2.8.1. Frustrating.

**Comment by quantumsheep** [comment]
Same thing happening, it works one time but not again. Restarting XQuartz fixes the issues but I have to restart it each time I'm using X11 from inside a Docker container.

**Comment by rajeevb2** [comment]
I am using macOS 12.1 and Ubuntu 20.04 and XQuartz 2.8.1 and I continue to get the following error when I start firefox (version 96) on Ubuntu.

libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast

Is there a patch for this I can apply or a workaround?

**Comment by ljexplore** [comment]
> I am using macOS 12.1 and Ubuntu 20.04 and XQuartz 2.8.1 and I continue to get the following error when I start firefox (version 96) on Ubuntu.
> 
> libGL error: No matching fbConfigs or visuals found libGL error: failed to load driver: swrast
> 
> Is there a patch for this I can apply or a workaround?

The workaround is above, but set the LIBGL_ALWAYS_INDIRECT environment variable to 1 on the Linux side

**Comment by as2875** [comment]
After upgrading to macOS 12.1 the workaround above (setting `enable_iglx`, restarting XQuartz) stopped working. I have tried setting LIBGL_ALWAYS_INDIRECT to 1.

I'm running XQuartz 2.8.1 and Ubuntu 18.04.

**Comment by raoultrines** [comment]
You should also make sure that the X11 server on the Mac is started with the flag +iglx enabled; this is disabled by default. See #231.

**Comment by jloera00** [comment]
I'm running XQuartz 2.8.2 and macOS 12.4.

When attempting to open a bxsf file on xcrysden and visualize the Fermi surface the following error occurs: 

```
xp_import_surface: assertion failed: s == NULL
error: xp_import_surface returned: 2
xp_destroy_surface: assertion failed: s != NULL
xp_destroy_surface error: 3
X Error of failed request: GLXBadContext
    Major opcode of failed request: 149
    ResourceID in failed request: 0x0
    Serial number of failed request: 13061
X Error of failed request: GLXBadCurrentWindow
    Major opcode of failed request: 149
    ResourceID in failed request: 0x0
    Serial number of failed request: 13061
```

Is there a suggested fix for this error? I tried `enable_iglx` and this did not help. glxgears show up just fine.

**Comment by ashutoshbhatt-cpi** [comment]
Any work-around solution for this issue?

**Comment by dillera** [comment]
Use Mobaterm on windows! It works fine.

**Comment by ashutoshbhatt-cpi** [comment]
@dillera: Yes, the X11 server (Xming) on Windows works fine. The problem is with XQuartz setup in Mac OS X

**Comment by sdykae** [comment]
any workarrounds?

**Comment by sdykae** [comment]
In my case osX Monterey workaround with firefox `ssh -XC user@ip`
- Configure the target: #231
`export LIBGL_ALWAYS_INDIRECT=1` on target linux in `.bashrc`
- Configure the host:
`defaults write org.xquartz.X11 enable_iglx -bool YES` on my mac osx monterey

**Comment by shaul-pollak** [comment]
any updates on this?

**Comment by woensug-choi** [comment]
[Setting Podman machine with X86_64 Fedora](https://developer.ibm.com/tutorials/running-x86-64-containers-mac-silicon-m1/?mhsrc=ibmsearch_a&mhq=x86%26comma%3B) and running Ubuntu 22.04 on M1 Macbook (OS Ventura) never succeded. But, running [tbillah's dockerfile](https://github.com/tashrifbillah/glxgears-containers/blob/master/Dockerfile) ran glxgears just fine! It's using `NVIDIA-Linux-x86_64-430.40.run` driver installation although the container has nothing much to do with NVIDIA card. I will stick to the version.

**Comment by pineapps** [comment]
```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
```

This issue is still occurring with XQuartz 2.8.5. It was working fine until I updated to macOS 13.2.1.

**Comment by PymZoR** [comment]
> ```
> libGL error: No matching fbConfigs or visuals found
> libGL error: failed to load driver: swrast
> ```
> 
> This issue is still occurring with XQuartz 2.8.5. It was working fine until I updated to macOS 13.2.1.
> 
> IGLX is enabled:
> 
> ```
> 
>     "NSWindow Frame SUStatusFrame" = "1720 1080 400 135 0 0 3840 1575 ";
>     "NSWindow Frame SUUpdateAlert" = "1610 882 620 398 0 0 3840 1575 ";
>     "NSWindow Frame SUUpdateAlertFrame" = "1632 882 575 398 0 0 3840 1575 ";
>     "NSWindow Frame x11_apps" = "698 639 454 299 0 0 3840 1575 ";
>     "NSWindow Frame x11_prefs" = "426 383 484 336 0 0 3840 1575 ";
>     SUHasLaunchedBefore = 1;
>     SULastCheckTime = "2023-03-19 20:42:12 +0000";
>     SUUpdateRelaunchingMarker = 0;
>     "app_to_run" = "/opt/X11/bin/xterm";
>     "apps_menu" =     (
>                 (
>             Terminal,
>             xterm,
>             n
>         )
>     );
>     "cache_fonts" = 1;
>     "done_xinit_check" = 1;
>     "enable_iglx" = 1;
>     "login_shell" = "/bin/sh";
>     "no_auth" = 0;
>     "nolisten_tcp" = 1;
>     "startx_script" = "/opt/X11/bin/startx -- /opt/X11/bin/Xquartz";
> ```

Same here !

**Comment by nbriggs** [comment]
I was having the same results after an OS update plus XQuartz update to 2.8.5, and I'm not sure which broke things, but checking and setting enable_iglx was the answer:

```
% defaults read org.xquartz.X11 enable_iglx
0
% defaults write org.xquartz.X11 enable_iglx -bool YES
% defaults read org.xquartz.X11 enable_iglx           
1
```
(and then make sure XQuartz is terminated and restarted)

The libGL errors for "No matching fbConfigs or visuals found" and "failed to load driver: swrast" from the remote client are still present, but, as far as I can tell, unrelated to the IGLX problem.

As Jeremy has said, check `~/Library/Logs/X11/org.xquartz.log` (and the `org.xquartz.log.old` if present) -
Prior to the `defaults write ...`:
```
X11.app: do_start_x11_server(): argc=7
        argv[0] = /opt/X11/bin/Xquartz
        argv[1] = :0
        argv[2] = -nolisten
        argv[3] = tcp
        argv[4] = -iglx
        argv[5] = -auth
        argv[6] = ...
```
Post `defaults write ...`:
```
X11.app: do_start_x11_server(): argc=9
        argv[0] = /opt/X11/bin/Xquartz
        argv[1] = :0
        argv[2] = -nolisten
        argv[3] = tcp
        argv[4] = +iglx
        argv[5] = +extension
        argv[6] = GLX
        argv[7] = -auth
        argv[8] = ...
```
Currently running XQuartz 2.8.5 (xorg-server 21.1.6) on macOS Big Sur 11.7.6 (20G1231)

**Comment by pprindeville** [comment]
I'm on a MacBook Pro (M1 Pro, 2021) running Ventura 13.4 and I have XQuartz 2.8.5 installed.

I get:

```
pprindeville@Philips-Laptop ~ % ssh -f -C -YAX2 kvm1  glxinfo
pprindeville@Philips-Laptop ~ % libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
name of display: localhost:10.0
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  149 (GLX)
  Minor opcode of failed request:  24 (X_GLXCreateNewContext)
  Value in failed request:  0x0
  Serial number of failed request:  26
  Current serial number in output stream:  27

```

Same result on a MacBook Air (2020, 14", Intel) also running 13.4 and XQuartz 2.8.5.

**Comment by nbriggs** [comment]
@pprindeville what's the result of your doing:
``` sh
% defaults read org.xquartz.X11 enable_iglx
1
```

**Comment by pprindeville** [comment]
I'm assuming I need to quit and restart XQuartz?  No change that I can tell.

**Comment by nbriggs** [comment]
@pprindeville well, no, not if the result of doing the `defaults read ...` is "1".  If you got "0" instead of "1", then you need to set it to true (1), by doing:
```
defaults write org.xquartz.X11 enable_iglx -bool YES
```
and then quit/restart XQuartz

**Comment by pprindeville** [comment]
Got:

```
pprindeville@Philips-Laptop ~ % defaults read org.xquartz.X11 enable_iglx
1
pprindeville@Philips-Laptop ~ % 
```
and:

```
/opt/X11/bin/Xquartz :0 -listen tcp +iglx +extension GLX -auth /Users/pprindeville/.serverauth.33107
```

**Comment by nbriggs** [comment]
@pprindeville well, that's the limit of my guessing at the problem.  I found that when I had upgraded between XQuartz versions the enable_iglx had somehow been reset to FALSE, and setting it back to TRUE solved the issue, and running X apps via an ssh session worked again.

**Comment by pprindeville** [comment]
Went to 2.8.4 and I'm seeing this there, too...

**Comment by pprindeville** [comment]
I guess there's little hope of this being fixed in 2.8.6...

**Comment by jeremyhu** [comment]
Yeah, not likely to be addressed ever unless someone comes along to maintain it, sorry...

**Comment by pprindeville** [comment]
> Yeah, not likely to be addressed ever unless someone comes along to maintain it, sorry...

Let me see if it’s in our budget to offer a bounty…

**Comment by kraken-jim** [comment]
Thanks to @nbriggs for:

> @pprindeville well, no, not if the result of doing the `defaults read ...` is "1". If you got "0" instead of "1", then you need to set it to true (1), by doing:
> 
> ```
> defaults write org.xquartz.X11 enable_iglx -bool YES
> ```
> 
> and then quit/restart XQuartz

**Comment by pprindeville** [comment]
> Thanks to @nbriggs for:
> 
> > @pprindeville well, no, not if the result of doing the `defaults read ...` is "1". If you got "0" instead of "1", then you need to set it to true (1), by doing:
> > ```
> > defaults write org.xquartz.X11 enable_iglx -bool YES
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and then quit/restart XQuartz

Did this and it makes no difference.

**Comment by pprindeville** [comment]
We're willing to negotiate a bounty to get this fixed, so if you're interested, toss out a number...

**Comment by pineapps** [comment]
> We're willing to negotiate a bounty to get this fixed, so if you're interested, toss out a number...

I'm interested and I will take a look. Out of curiosity, has anyone tried the 2.8.6 beta release? https://www.xquartz.org/releases/XQuartz-2.8.6_beta2.html

**Comment by nbriggs** [comment]
@pineapps - about to try the 2.8.6_beta2 release (in my case, to make sure there have been regressions in things I care about).

The link on the release page is wrong, though, but one can easily guess the correct one:  https://github.com/XQuartz/XQuartz/releases/download/XQuartz-2.8.6_beta1/XQuartz-2.8.6_beta2.pkg should be `https://github.com/XQuartz/XQuartz/releases/download/XQuartz-2.8.6_beta2/XQuartz-2.8.6_beta2.pkg

**Comment by s-daveb** [comment]
Using the 2.8.6_beta2 release, and setting LIBGL_ALWAYS_INDIRECT=1 gets rid of any errors, but when the glxgears window appears on-screen, it does not animate, and the program will sometimes output a ridiculous framerate to terminal, or nothing at all.

**Comment by nbriggs** [comment]
@s-daveb - my glxgears, talking to XQuartz 2.8.6_beta2, animates very very slowly, but it does move.  It reports:
```
% glxgears
11711 frames in 5.8 seconds = 2020.886 FPS
389 frames in 6.5 seconds = 59.839 FPS
388 frames in 6.5 seconds = 59.690 FPS
388 frames in 6.5 seconds = 59.693 FPS
389 frames in 5.2 seconds = 75.301 FPS
305 frames in 6.3 seconds = 48.655 FPS
389 frames in 6.5 seconds = 59.835 FPS
```
(and that glxgears is running on an ancient 386 box under FreeBSD).  The first frame-rate output is always ridiculous, but after that it pretty much settles down to ~60 FPS.

**Comment by ljexplore** [comment]
@s-daveb yes, I see the same behavior. It appears that the gears are actually being redrawn extremely slowly, I can see individual pixels changing... and I get a similar printout as @nbriggs:

> glxgears
30794 frames in 21.8 seconds = 1409.476 FPS
1404 frames in 8.7 seconds = 161.985 FPS
527 frames in 22.8 seconds = 23.096 FPS
1304 frames in 7.9 seconds = 164.384 FPS
1191 frames in 23.9 seconds = 49.834 FPS
1365 frames in 8.7 seconds = 157.163 FPS
532 frames in 8.7 seconds = 61.381 FPS
533 frames in 8.7 seconds = 61.497 FPS
529 frames in 8.7 seconds = 61.030 FPS

Even though glxgears behaves oddly, interactive OpenGL programs appear to operate normally.

I'll note that currently ssh-ing into an Ubuntu system running Ubuntu 20.04.6 I need to set LIBGL_ALWAYS_INDIRECT to get rid of the error, whereas ssh-ing into a RHEL 7.9 system I don't, and glxgears runs about as expected (this is with macOS 12.6.7 running Quartz 2.8.2 locally as mentioned). 

NVIDIA GL libs are installed on both systems, but on the Ubuntu system, libGLX_indirect.so.0 is linked to libGLX_mesa.so.0, whereas on the RHEL7 system it's linked to libGLX_nvidia.so.535.54.03 so that may be the cause of the difference in behavior.

**Comment by lina196** [comment]
I have MacBook pro M1, so there are some missing graphics. I downloaded those packages, and then the GUI worked fine without error:  
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast

Those or the links:
https://docs.mesa3d.org/macos.html
https://docs.mesa3d.org/download.html

**Comment by Luke20000429** [comment]
Try install XQuartz 2.8.6 beta2 and `export LIBGL_ALWAYS_INDIRECT=1`, still getting `Error: glXCreateContext failed`.

**Comment by vpuri3** [comment]
Is it possible to do direct rendering? Would that help sidestep the `swrast` issue?

```bash
[user@remoteMachine ~]:glxinfo | grep "direct rendering"
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
direct rendering: No (If you want to find out why, try setting LIBGL_DEBUG=verbose)
X Error of failed request:  GLXBadContextTag
  Major opcode of failed request:  150 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  54
  Current serial number in output stream:  54
```

**Comment by ryan-utopia** [comment]
Since XQuartz doesn't support GLX on Mac(M Chips), this may be a sulotion:
Use tools like xvfb, x11vnc, noVNC, and xfce4 to support GUI applications such as RViz and Gazebo.

See This docker [Docker](https://hub.docker.com/repository/docker/ryanchennn/arm64-ros-noetic-vnc/general)

**Comment by kmatheussen** [comment]
You'll probably get much better GLX performance running a Linux VM in
UTM than using VNC.
The graphical performance of UTM is pretty extreme actually, but it
doesn't have the same type of full screen that XQuartz has. UTM has
that extremely annoying delay when switching between MacOS and UTM
(XQuartz is the only full-screen program I've seen which doesn't have
that), plus that UTM renders everything below the notch so you can't
exploit all of the screen.

**Comment by benz0li** [comment]
> You'll probably get much better GLX performance running a Linux VM in UTM than using VNC.

Cross reference:

* https://github.com/XQuartz/XQuartz/issues/365#issuecomment-1762918977

**Comment by kmatheussen** [comment]
Yes, and It's important to point out that debian and ubuntu are currently
unstable with UTM.
It freezes after a while. Use Fedora instead.

**Comment by maehne** [comment]
The issue persists and prevents starting GUI applications using OpenGL from inside a container. E.g., we are using the [IIC-OSIC-TOOLS](https://github.com/iic-jku/IIC-OSIC-TOOLS) container, which packages many open-source electronic design automation tools necessary for ASIC design using open-source PDKs. Normal X11 applications, launched from within the container, are displayed properly by `X11.app`. However, applications depending on OpenGL consistently fail, e.g., `GDS3D` (see https://github.com/iic-jku/IIC-OSIC-TOOLS/issues/220) with:

```
No matching fbConfigs or visuals found
glx: failed to create drisw screen
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  148 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  37
  Current serial number in output stream:  37
```

`glxinfo` fails with:
```
$ LIBGL_ALWAYS_INDIRECT=1 glxinfo
name of display: host.docker.internal:0
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  148 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  24
  Current serial number in output stream:  24
$ LIBGL_ALWAYS_INDIRECT=0 glxinfo
name of display: host.docker.internal:0
No matching fbConfigs or visuals found
glx: failed to create drisw screen
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  148 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  33
  Current serial number in output stream:  33
```

`glxgears` fails with:
```
$ LIBGL_ALWAYS_INDIRECT=1 glxgears
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  148 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  26
  Current serial number in output stream:  26
$ LIBGL_ALWAYS_INDIRECT=0 glxgears
No matching fbConfigs or visuals found
glx: failed to create drisw screen
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  148 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  35
  Current serial number in output stream:  35
```

The output is similar with the local `glxinfo` and `glxgears` when forcing indirect rendering:
```
$ LIBGL_ALWAYS_INDIRECT=1 glxgears
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  148 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  24
  Current serial number in output stream:  24
$ LIBGL_ALWAYS_INDIRECT=1 glxinfo
name of display: /var/run/com.apple.launchd.kD9nBh9xYl/org.macports:0
X Error of failed request:  GLXBadContext
  Major opcode of failed request:  148 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  22
  Current serial number in output stream:  22
```

XQuartz and its dependencies have been installed with MacPorts on my up-to-date macOS 26.5 on a MacBook Air M4:
```
$ port info xorg-server
xorg-server @1.20.11_1 (x11, devel)
[...]
```

**Comment by jeremyhu** [comment]
This might be a bit improved with 2.8.7_beta1.  I was focusing a bit more on the direct rendering cases, but some of the work I did would impact indirect rendering too...

**Comment by jeremyhu** [comment]
I'm able to get basic rendering (eg: glxgears) working fine over indirect GLX with 2.8.7_beta2, so closing this out as "improved"... but please file new tickets with specific failures to dig into as followups.
